### PR TITLE
Task2 from networking is ready

### PR DESCRIPTION
**Pull request by qnislavnachev**

### CODE REVIEW

File: src/task2/server/Server.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            System.out.println("Server is set up and waiting for users...");
+            while (true) {
+                connection = server.accept();
+                System.out.println("User log into server!");
+                PrintStream printer = new PrintStream(connection.getOutputStream());
+                printer.println("Hello from the server ! " + date);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        } finally {
```
Comment (tsony-tsonev) [inline src/task2/server/Server.java]: same here

File: test/clientTest/FakeServer.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package clientTest;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+
+public class FakeServer {
```
Comment (tsony-tsonev) [inline test/clientTest/FakeServer.java]: your tests should not depend on starting a server before running the tests, you should start the fake server inside the before method in the client tests

File: test/serverTest/RunServer.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package serverTest;
+
+import task2.server.Server;
+
+public class RunServer {
```
Comment (tsony-tsonev) [inline test/serverTest/RunServer.java]: same here

File: test/clientTest/FakeServer.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    public static void main(String[] args) throws IOException {
+        ServerSocket fakeServer = null;
+        fakeServer = new ServerSocket(1111);
+        Socket connection = fakeServer.accept();
+        PrintStream printer = new PrintStream(connection.getOutputStream());
+        printer.println("Hello");
+        fakeServer.close();
+        connection.close();
```
Comment (tsony-tsonev) [inline test/clientTest/FakeServer.java]: this is socket not connection

File: test/clientTest/TestClient.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.*;
+public class TestClient {
+
+    @Test
+    public void ClientReceiveMessage() throws Exception {
+        Client client = new Client();
+        client.join("localhost", 1111);
+        String expected = client.getMessage();
```
Comment (tsony-tsonev) [inline test/clientTest/TestClient.java]: this is the actual

File: test/serverTest/TestServer.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.*;
+
+public class TestServer {
+
+    @Test
+    public void serverSendMessage() throws Exception {
+        Socket fakeClient = new Socket("localhost", 1111);
+        BufferedReader reader = new BufferedReader(new InputStreamReader(fakeClient.getInputStream()));
+        String expected = reader.readLine();
```
Comment (tsony-tsonev) [inline test/serverTest/TestServer.java]: this is the actual

File: src/task2/client/Client.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.Socket;
+
+public class Client {
+    private String message = null;
+
+    public void join(String host, int port) {
+        Socket client = null;
+        try {
```
Comment (qnislavnachev) [inline src/task2/client/Client.java]: fixed

File: src/task2/server/Server.java (hunk 1 above)
Comment (qnislavnachev) [inline src/task2/server/Server.java]: fixed

File: test/clientTest/FakeServer.java (hunk 2 above)
Comment (qnislavnachev) [inline test/clientTest/FakeServer.java]: fixed

File: test/serverTest/RunServer.java (hunk 3 above)
Comment (qnislavnachev) [inline test/serverTest/RunServer.java]: fixed

File: test/clientTest/TestClient.java (hunk 5 above)
Comment (qnislavnachev) [inline test/clientTest/TestClient.java]: fixed

File: test/serverTest/TestServer.java (hunk 6 above)
Comment (qnislavnachev) [inline test/serverTest/TestServer.java]: fixed

File: test/serverTest/TestServer.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            }
+        });
+        clientThread.start();
+        server.start();
+    }
+
+    @Test
+    public void serverSendMessage() throws Exception {
+        String actual = fakeClient.getMessage();
+        String expected = "Hello from the server ! Wed Sep 28 17:09:46 EEST 2016";
```
Comment (tsony-tsonev) [inline test/serverTest/TestServer.java]: are you sure that this test is working?

Comment (qnislavnachev) [inline test/serverTest/TestServer.java]: fixed

Comment (tsony-tsonev) [inline test/serverTest/TestServer.java]: Nope, the problem is that you call new Date() in the production code and in the test, so the two dates are never equal (by milliseconds). You should pass to your server some interface which has a method that gives you the date, this way in the test you can easily fake that interface and make it return the same date every time you ask it. The test will then pass with no problems.

Comment (tsony-tsonev) [inline test/serverTest/TestServer.java]: You should inject the interface which will give you the date in the constructor of the server

File: src/task2/server/ServerDate.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package task2.server;
+
+import java.util.Date;
+
+public interface ServerDate {
```
Comment (tsony-tsonev) [inline src/task2/server/ServerDate.java]: Ok good, that was the idea, but what if you want to reuse it in the client or somewhere else you will be forced to create new interface ClientDate and they both will be the same except for the names. Interfaces should have common names so they can be reused think of something better, something related to date and not coupled with the name server.

**Comment by tsony-tsonev** [comment]
LGTM afrer fixing the last name

**Pull request merged**